peer.go

```diff
@@ -74,81 +83,76 @@ func (l *PeerList) newSibling() *PeerList {
 
 // Add adds a peer to the list if it does not exist, or returns any existing peer.
 func (l *PeerList) Add(hostPort string) *Peer {
-	l.RLock()
-
-	if p, ok := l.peersByHostPort[hostPort]; ok {
-		l.RUnlock()
-		return p
+	if ps, ok := l.exists(hostPort); ok {
+		return ps.Peer
 	}
-
-	l.RUnlock()
 	l.Lock()
 	defer l.Unlock()
 
 	if p, ok := l.peersByHostPort[hostPort]; ok {
-		return p
+		return p.Peer
 	}
 
 	p := l.parent.Add(hostPort)
-	atomic.AddUint32(&p.scCount, 1)
-	l.peersByHostPort[hostPort] = p
-	l.peers = append(l.peers, p)
+	ps := newPeerScore(p)
+	ps.score = l.scoreCalculator.GetScore(p)
```

Comment:
put the score calculation directly in side of `newPeerScore` rather than giving it some initial score and replacing it immediately